docs/zephyr_02_functional_requirements.sdoc

```diff
@@ -1379,6 +1364,120 @@ As a Zephyr OS user I want to... ???
 
 [/SECTION]
 
+[SECTION]
+TITLE: Thread Synchronization
+
+[REQUIREMENT]
+UID: ZEP-990001
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Semaphore
+TITLE: Counting Semaphore Creation
+STATEMENT: >>>
+The system shall provide a mechanism to create semaphores, both statically at build time and dynamically at runtime, ensuring deterministic operation with bounded execution times.
+<<<
+USER_STORY: >>>
+As a software developer tasked with implementing resource management in our real-time operating system (RTOS), I want to integrate a semaphore synchronization primitive to facilitate coordinated access to shared resources among multiple tasks.
+<<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-99
+
+[REQUIREMENT]
+UID: ZEP-990002
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Semaphore
+TITLE: Semaphore Initialization
+STATEMENT: >>>
+The system shall provide a mechanism to initialize semaphore values to regulate the number of tasks or resources concurrently accessing a shared resource.
+<<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-99
+
+[REQUIREMENT]
+UID: ZEP-990010
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Semaphore
+TITLE: Semaphore Initialization
+STATEMENT: >>>
+When initializing a counting semaphore, the system shall accept options specifying initial semaphore value and the maximum permitted count a semaphore can have.
+<<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-99
+
+[REQUIREMENT]
+UID: ZEP-990003
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Semaphore
+TITLE: Semaphore acquisition and release
+STATEMENT: >>>
+The system shall provide a mechanism enabling tasks to acquire (take) and release (give) semaphores, ensuring atomic and thread-safe operations for semaphore acquisition and release.
+<<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-99
+
+[REQUIREMENT]
+UID: ZEP-990004
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Semaphore
+TITLE: Semaphore Counting
+STATEMENT: >>>
+The system shall support counting semaphores that allow multiple tasks to increment and decrement counts to manage resource availability or access permissions.
+<<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-99
+
+[REQUIREMENT]
+UID: ZEP-990006
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Semaphore
+TITLE: Semaphore Counting
+STATEMENT: >>>
+When counting semaphores is in use, the system shall accurately track resource usage and prevent overflows or underflows.
```

Comment:
**question:** When stated as is, what is the (implicit) implication of When counting semaphores are not in use? What mustn't overflow or or underflow?